runtime/callback.c

```diff
@@ -256,35 +256,29 @@ CAMLexport value caml_callback3_exn(value closure,
   }
 }
 
-CAMLexport value caml_callbackN_exn(value closure, int narg, value args[])
-{
-  CAMLparam1 (closure);
-  CAMLxparamN (args, narg);
-  CAMLlocal1 (res);
-  int i;
-
-  res = closure;
-  for (i = 0; i < narg; /*nothing*/) {
-    /* Pass as many arguments as possible */
-    switch (narg - i) {
+CAMLexport value caml_callbackN_exn(value closure, int narg, value args[]) {
+  switch (narg) {
+    case 0:
+      return closure;
```

Comment:
Note: the previous implementation gracefully handles the case `narg == 0`, so I kept this property in the new implementation. I suspect that it is not handled well in the bytecode implementation, so presumably there is an implicit precondition that `narg > 0`. We could clarify this condition, or ensure that the bytecode version also works well with `narg == 0`, I went with keeping the behavior unchanged in both versions.